cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/trigger/AbstractCompositeTrigger.java

```diff
@@ -70,15 +74,30 @@ public void updateLaunchArguments(ProgramSchedule schedule, List<Notification> n
     return unitTriggers;
   }
 
+  /**
+   * Get all inner triggers as SatisfiableTrigger's in this trigger. Cannot initialize satisfiableTriggers in
+   * the constructor because the constructor will not be called if this is instantiated by {@link TriggerCodec}.
+   */
+  public List<SatisfiableTrigger> getSatisfiableTriggers() {
+    if (satisfiableTriggers == null) {
+      satisfiableTriggers = new ArrayList<>();
+      for (Trigger trigger : getTriggers()) {
+        // the trigger is guaranteed to be an instance of SatisfiableTrigger because of the parameter type of
+        // the constructor and the mapping in TriggerCodec
+        satisfiableTriggers.add((SatisfiableTrigger) trigger);
+      }
+    }
+    return satisfiableTriggers;
+  }
+
   private void initializeUnitTriggers() {
     unitTriggers = new HashMap<>();
     for (Trigger trigger : getTriggers()) {
       // Add current non-composite trigger to the corresponding set in the map
-      Type triggerType = ((ProtoTrigger) trigger).getType();
-      if (trigger instanceof co.cask.cdap.internal.app.runtime.schedule.trigger.AbstractCompositeTrigger) {
+      Type triggerType = trigger.getType();
+      if (trigger instanceof AbstractCompositeTrigger) {
         // If the current trigger is a composite trigger, add each of its unit triggers to the set according to type
-        for (Map.Entry<Type, Set<SatisfiableTrigger>> entry :
-          ((co.cask.cdap.internal.app.runtime.schedule.trigger.AbstractCompositeTrigger) trigger)
+        for (Map.Entry<Type, Set<SatisfiableTrigger>> entry : ((AbstractCompositeTrigger) trigger)
```

Comment:
this is hard to read. Cast it before the for loop or move `((AbstractCompositeTrigger) trigger)` to the next line